### PR TITLE
Modification Request

### LINKED ISSUE #37: Profile Modification Request
As an external user I can,
- Initiate request to edit my own account

### LINKED ISSUE #61: Request Modification Employee and Manager
As an employee or manager I can,
- View all modification requests of external user
- View a single modification request of external user
- Approve or reject request of external user
- Initiate request to edit my own account

### LINKED ISSUE #62: Modification Request for Admin
As an admin I can,
- View all modification requests of internal user
- View a single modification request of internal user
- Approve or reject request of internal user

### PR DESCRIPTION
**Pull request by ayush2k**
Resolves #37, Resolves #61, Resolves #62 
Description:
- External User can initiate modification request for profile
- Internal Users can view and approve or reject requests
- Admin can approve or reject internal user requests

### CODE REVIEW

File: src/main/java/securbank/controller/AdminController.java [hunk 1]
```diff
@@ -111,9 +108,69 @@ public String getUserDetail(Model model, @PathVariable UUID id) {
...
+		logger.info("GET request: User modification request by ID");
+		
+		
+        return "admin/modificationrequest_detail";
+    }
+	
+	@PostMapping("/admin/user/request/{requestId}")
+    public String approveEdit(@PathVariable UUID requestId, @ModelAttribute ModificationRequest request, BindingResult bindingResult) {
+		String status = request.getStatus();
+		if (status == null || !(request.getStatus().equals("approved") || !request.getStatus().equals("rejected"))) {
```
Comment (Nikh13) [inline src/main/java/securbank/controller/AdminController.java]: When you reject a request it gives an error. The logic isn't right here. The ! outside is fine but there shouldn't be another one inside for 'rejected'.

File: src/main/java/securbank/controller/EmployeeController.java [hunk 2]
```diff
@@ -19,18 +29,105 @@
...
+		model.addAttribute("modificationrequest", modificationRequest);
+		logger.info("GET request: Employee external modification request by ID");
+		
+        return "employee/modificationrequest_detail";
+    }
+	
+	@PostMapping("/employee/user/request/{requestId}")
+    public String approveEdit(@PathVariable UUID requestId, @ModelAttribute ModificationRequest request, BindingResult bindingResult) {
+		String status = request.getStatus();
+		if (status == null || !(request.getStatus().equals("approved") || !request.getStatus().equals("rejected"))) {
```
Comment (Nikh13) [inline src/main/java/securbank/controller/EmployeeController.java]: Same bug as admincontroller.

File: src/main/java/securbank/controller/ManagerController.java [hunk 3]
```diff
@@ -41,6 +51,90 @@ public String currentUserDetails(Model model) {
...
+		model.addAttribute("modificationrequest", modificationRequest);
+		logger.info("GET request: Manager external modification request by ID");
+		
+        return "manager/modificationrequest_detail";
+    }
+	
+	@PostMapping("/manager/user/request/{requestId}")
+    public String approveEdit(@PathVariable UUID requestId, @ModelAttribute ModificationRequest request, BindingResult bindingResult) {
+		String status = request.getStatus();
+		if (status == null || !(request.getStatus().equals("approved") || !request.getStatus().equals("rejected"))) {
```
Comment (Nikh13) [inline src/main/java/securbank/controller/ManagerController.java]: Same bug as admincontroller.

File: src/main/resources/templates/employee/detail.html [hunk 4]
```diff
@@ -5,18 +5,18 @@
...
+	        <p>Email: <p th:text="*{email}" /></p>
+	        <p>First Name: <p th:text="*{firstName}" /></p>
+	        <p>Last Name: <p th:text="*{lastName}" /></p>
+	        <p>Phone: <p th:text="*{phone}" /></p>
+	        <p>Address Line1: <p th:text="*{addressLine1}" /></p>
+	        <p>City: <p th:text="*{city}" /></p>
+	        <p>Zip: <p th:text="*{zip}" /></p>
+	        <p>State: <p th:text="*{state}" /></p>
+	        <p>Role: <p th:text="*{role}" /></p>
+			<p><a th:href="@{employee/edit}">Edit</a></p>
```
Comment (Nikh13) [inline src/main/resources/templates/employee/detail.html]: This should be /employee/edit

File: src/main/resources/templates/manager/detail.html [hunk 5]
```diff
@@ -2,7 +2,7 @@
 <html xmlns="http://www.w3.org/1999/xhtml" xmlns:th="http://www.thymeleaf.org"
       xmlns:sec="http://www.thymeleaf.org/thymeleaf-extras-springsecurity3">
     <head>
-        <title>Spring Security Example </title>
+        <title>Details</title>
```
Comment (Nikh13) [inline src/main/resources/templates/manager/detail.html]: Similar to the employee details page, the link to the edit page should be /manager/edit and needs to be changed for external user as well.

**Review by Nikh13 — CHANGES_REQUESTED** [review]
Few minor changes. Let me know when theyre done. Ill merge.

File: src/main/java/securbank/services/UserServiceImpl.java [hunk 6]
```diff
@@ -250,4 +288,315 @@ public NewUserRequest getNewUserRequest(UUID newUserRequestId) {
...
+		ModificationRequest request = new ModificationRequest();
+		List<ModificationRequest> requests = modificationRequestDao.findAllbyUser(current);
+		
+		// Deactivate all active request 
+		for (ModificationRequest activeRequest : requests) {
+			activeRequest.setActive(false);
+			modificationRequestDao.update(activeRequest);
+		}
+		
+		if (!request.getEmail().equals(current.getEmail())) {
```
Comment (Nikh13) [inline src/main/java/securbank/services/UserServiceImpl.java]: Error thrown at this line. I think it should be user.getEmail() and not request.getEmail().

**Review by Nikh13 — CHANGES_REQUESTED** [review]

File: src/main/resources/templates/manager/edit.html [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+			Zip: <input type="text" th:field="*{zip}" />
+		<p th:if="${#fields.hasErrors('zip')}" th:errors="*{zip}"></p>
+		</p>
+		<p>
+			State: <input type="text" th:field="*{state}" />
+		<p th:if="${#fields.hasErrors('state')}" th:errors="*{state}"></p>
+		</p>
+		<p>
+			Role: <select th:field="*{role}">
+				<option th:value="ROLE_EMPLOYEE" th:selected="(*{role} == ROLE_EMPLOYEE")>Employee</option>
```
Comment (Nikh13) [inline src/main/resources/templates/manager/edit.html]: The double quotes are ending before ')' in the th:selected

**Review by Nikh13 — CHANGES_REQUESTED** [review]

File: src/main/java/securbank/controller/ManagerController.java (hunk 3 above)
Comment (Nikh13) [reply src/main/java/securbank/controller/ManagerController.java]: You've fixed the bug in all the other files but I think you missed it over here.

File: src/main/java/securbank/services/UserServiceImpl.java (hunk 6 above)
Comment (ayush2k) [reply src/main/java/securbank/services/UserServiceImpl.java]: @Nikh13 `request.getEmail()` is working for me. Email of request is set in the preceding lines. Though the we didn't have the mapping for verifying the email. 
Please confirm if this is correct.
Thanks

File: src/main/resources/templates/manager/edit.html (hunk 7 above)
Comment (ayush2k) [reply src/main/resources/templates/manager/edit.html]: Fixed

**Review by Nikh13 — APPROVED** [review]

**Pull request merged**